### DISCUSSION TITLE
[RFC] New Component: ColorPicker 拾色器

### DISCUSSION THREAD

**Discussion by MadCcc** [ant-design/ant-design#40974] [RFCs - archive]
<a name="Summary"></a>
## Summary
Provide new component ColorPicker in later minors of 5.x.
<a name="Motivation"></a>
## Motivation
For a long time, ColorPicker with style of antd has been proposed in community, such as [https://github.com/ant-design/ant-design/issues/35315](https://github.com/ant-design/ant-design/issues/35315) . However, we didn't introduce it due to its complex implementation and lack of design.<br />But now we want to bring it into antd 5.x version. It's time to have these issues resolved. 
<a name="API"></a>
## API
| **Property** | **Description** | **Type** | **Default** |
| --- | --- | --- | --- |
| format | Format of color. | `rgb` &#124; `hex` &#124; `hsb` | `hex` |
| onFormatChange | Callback when `format` is changed. | `(format: 'hex' \| 'rgb' \| 'hsb') => void` | - |
| value | Value of color. | string &#124; `Color` | - |
| defaultValue | Default value of color. | string &#124; `Color` | - |
| onChange | Callback when `value` is changed. | `(value: Color, hex: string) => void` | - |
| allowClear | Allow clearing color selected. | boolean | false |
| presets | Preset colors. | { label: ReactNode, colors: Array<string &#124; Color> }[] | - |
| layout*(always vertical) | Layout of popup. | `horizontal` &#124; `vertical` | `vertical` |
| children | Trigger of ColorPicker. | ReactElement | - |
| trigger | ColorPicker trigger mode. | `hover`&#124; `click` | `hover` |
| open | Whether to show popup. | boolean | - |
| onOpenChange | Callback when `open`is changed. | `(open: boolean) => void` | - |
| showArrow* | Whether to show the drop-down arrow | boolean | - |
| showValue* | Whether to show selected color | boolean | - |
| icon* | Custom icon in trigger. | ReactNode | - |
| disabled | Disable ColorPicker | boolean | - |
| placement | Placement of popup | `top` \| `topLeft` \| `topRight` \| `bottom` \| `bottomLeft` \| `bottomRight` | `bottomLeft` |
| arrow | Configuration for popup arrow | boolean \| `{ pointAtCenter: boolean }` | - |

> `className` and `style` is supported by default.
> properties with `*` are not necessary in first version.

<a name="9afe3fae"></a>
## Basic Example
<a name="de09e0ca"></a>
### Basic
![image](https://user-images.githubusercontent.com/27722486/221877634-de317f86-fd01-4c62-867f-7ac3c9656239.png)

```tsx
import React from 'react';
import { ColorPicker } from 'antd';

export default () => (
  <ColorPicker />
);
```
<a name="WZF7c"></a>
### Controlled
```tsx
import React from 'react';
import { ColorPicker } from 'antd';

export default () => {
  const [color, setColor] = useState<string>('#000');

  const handleChange = (newColor) => {
    setColor(newColor.hex);
  }

  return (
    <ColorPicker
      value={color}
      onChange={handleChange}
    />
  );
}
```
<a name="ideqW"></a>
### AllowClear
![image](https://user-images.githubusercontent.com/27722486/221877727-7a605a18-4b6e-48cb-a176-1252b412336b.png)

```tsx
import React from 'react';
import { ColorPicker } from 'antd';

export default () => (
  <ColorPicker allowClear />
);
```
<a name="mVtYI"></a>
### PresetColors
![image](https://user-images.githubusercontent.com/27722486/221877816-3a6646c4-7561-4296-bd58-c5732c25a3cc.png)
![image](https://user-images.githubusercontent.com/27722486/221877859-9eb5d230-84fe-4781-80d0-c47e1fb903d7.png)

```tsx
import React from 'react';
import { ColorPicker } from 'antd';

export default () => (
  <>
    <ColorPicker
      presets={[
        {
          label: 'Recommended',
          colors: ['#000', '#fff', '#163cff'],
        },
        {
          label: 'Recent',
          colors: [],
        }
      ]}
    />
    <ColorPicker
      allowClear
      layout="horizontal"
      presets={[
        {
          label: 'Recommended',
          colors: ['#000', '#fff', '#163cff'],
        },
        {
          label: 'Recent',
          colors: [],
        }
      ]}
    />
  </>
);
```
<a name="BRRbm"></a>
### Triggers
![image](https://user-images.githubusercontent.com/27722486/221877996-aba62ebd-fed1-4ca3-8f87-2ca4b8ed6cf9.png)

```tsx
import React from 'react';
import { ColorPicker, Button } from 'antd';
import { FontColorsOutlined } from '@ant-design/icons';

export default () => (
  <>
    <div>Show Arrow</div>
    <ColorPicker arrow/>
    <div>Show Value</div>
    <ColorPicker format="rgb" showValue />
    <div>Icon</div>
    <ColorPicker icon={<FontColorsOutlined />} />
    <div>Custom Trigger</div>
    <ColorPicker>
      <Button>Color</Button>
    </ColorPicker>
  <>
);
```
<a name="d24309fc"></a>
## Detailed Design
<a name="s9HRI"></a>
### Layout
![image](https://user-images.githubusercontent.com/27722486/221878127-80af6538-2b88-4951-8d1f-6a0faf4d8e4a.png)

<a name="xjf6p"></a>
### Color Panel
![image](https://user-images.githubusercontent.com/27722486/221878197-8de905c1-cea5-44a4-9e95-164a78e47b98.png)
<br />The first picker is responsive in vertical.
<a name="cqv2P"></a>
### Format
<a name="wiMqU"></a>
#### HEX
![image](https://user-images.githubusercontent.com/27722486/221878242-1e754eee-2f70-44b6-913d-d2addda0719d.png)

1. Add `#` prefix automaticly.
2. Transform to capitalized when blurred.
<a name="faGZb"></a>
#### RGB
![image](https://user-images.githubusercontent.com/27722486/221878293-77b17ffd-a298-494c-85a9-c0e934fb520a.png)

1. Input range is `0~255`.
2. `step`is `1`
<a name="hqFLC"></a>
#### HSB
![image](https://user-images.githubusercontent.com/27722486/221878404-43fcb60a-3830-440f-a361-3b56a916149e.png)

1. Input range for `H` is `0~360`, `step` is `1`
2. Input range for `S` and `B` is `0~100`, `step` is `1`.
3. `S`and `B` have suffix `%`
<a name="aZWUW"></a>
### Custom Format
In order to support more color formats, we need to have format part configurable. Currently ColorPicker do not support custom format, but this could be an internal feature. In order to implement it, format inputs should be components that have similar props as ColorPicker. And also, internal `color` should support transformation from any format provided.<br />So in we would implement ColorPanel and other feature in `@rc-component/color-picker`, then provide `hex` `rgb` `hsb` format input in `antd`.

### Internal Color
Use [@ctrl/tinycolor](https://github.com/scttcper/tinycolor) to transform input. 
Color Object should have following methods:
| method | description |
| --- | --- |
| toHexString | Transform to `hex` string starts with `#`. Could be `hex8` if has alpha. |
| toRgbString | Transform to `rgb` or `rgba` string. |
| toRgb* | Transform to `rgb` or `rgba` object. |
| toHsbString | Transform to `hsb` or `hsba` string. |
| toHsb* | Transform to `hsb` or `hsba` object. |

## How we teach this

1. Add new page for ColorPicker in document, and provide enough demos and api table.
2. Fill in the changelog.

### `rc-color-picker`
Implement `rc-color-picker` in https://github.com/react-component/color-picker . `rc-color-picker` should have the core features of antd ColorPicker, including:
- Color Panel
- Popup

Other feature could be implemented in `antd`. including:
- Presets
- Internal trigger
- Input of all color format
- styles

**Comment by arvinxx** [comment]
onChange 一个 props 不够。有两种时机，一种是 onStopChange，一种是 onDraggingChange，后者在实时刷新强的时候会更有用。前者一般用在性能要求高，或者中间态不重要的场景。

**Reply by MadCcc** [reply]
一开始会做的简单点，只有 `onStopChange`。这两种在非受控模式下行为不一样，也得区分。

**Comment by RedJue** [comment]
Mark
